cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/preview/PreviewHttpHandler.java

```diff
@@ -222,4 +298,37 @@ private ProgramRunId getProgramRunId(String namespaceId, String previewId) throw
   private RunRecordMeta getRunRecord(String namespaceId, String previewId) throws Exception {
     return previewManager.getRunner(new ApplicationId(namespaceId, previewId)).getRunRecord();
   }
+
+  private MetricsQueryHelper getMetricsQueryHelper(String namespaceId, String previewId) throws Exception {
+    return previewManager.getRunner(new ApplicationId(namespaceId, previewId)).getMetricsQueryHelper();
+  }
+
+  private List<String> overrideTags(List<String> tags, String namespaceId, String previewId) {
+    List<String> newTags = Lists.newArrayList();
+    String namespaceTag = NAMESPACE_STRING + ":" + namespaceId;
+    String previewTag = APP_STRING + ":" + previewId;
+    for (String tag : tags) {
+      if (tag.startsWith(NAMESPACE_STRING)) {
+        newTags.add(namespaceTag);
+      } else if (tag.startsWith(APP_STRING)) {
+        newTags.add(previewTag);
+      } else {
+        newTags.add(tag);
+      }
+    }
```

Comment:
so this handles the case where the namespace is provided in tags, but we need to add tags ourselves for namespace and app id if they are not provided as tags (as they are available in path parameter).